### PR TITLE
Issue #1217 - Domain expiration date - EPP

### LINKED ISSUE #1217: Implement setting registry expiration date via EPP
### Issue description

The registry has an expiration date but we are also storing the expiration date on the domain.py model. There  are property tagged functions for getting and setting the **registry** expiration date with only the getter implemented (though not tested). This should not be confused with the stored expiration date on the domain that appears in the model table. We need a way for the **registry** expiration date to updated from our registrar, thus we need to implement the setter portion. 

We do not need to provide a way on django admin for an analyst or user to update this expiration date as that is still under discussion as how best to do this. Right now NO one should be able to actually change the expiration date via the model

### Acceptance criteria

- [ ]  update registry expiration date getter as needed
- [ ] implement setting the registry expiration date 
- [ ] on django admin don't allow expiration date to be editable
- [ ] add unitest for the setter
- [ ] in the setter update the domain.expiration date with this new set date

### Additional context
Goal of this ticket is  to be used for the setting expiration on epp via a script in the blocked ticket. 

_No response_

### Links to other issues

blocked by: #1215 
blocks: #1212

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #1217

## Changes

- Updated django admin so that expiration date is read only
- Disabled expiration date setter (not implemented)
- Implemented renew domain, which sets the expiration date in the registry, registrar, and cache

## Context for reviewers

Test in [DK Sandbox](https://getgov-dk.app.cloud.gov/)

For **design review**, all that needs to be verified for this one is that the expiration date is readonly in django admin.  Navigate to a domain detail page in django admin.  Note that the expiration date is now readonly instead of editable.

For **developer code review**, note the following changes:
**admin.py**  set the expiration date to readonly to satisfy django admin requirement
**models/domain.py**  disabled the expiration date setter, as setting an exact date is not implemented at this time; implemented renew_domain which sets the expiration date to some unit of time in the future from the current expiration date (if no expiration date exists, sets from today); getter modified to also update registrar and cache when retrieving expiration date from registry

For **developer verification**, there is no UI for testing this.  So, you will need to go into sandbox command line and run a few commands, and note the output.
```
# connect to sandbox, and to python shell
cf ssh getgov-dk
/tmp/lifecycle/shell
./manage.py shell
```
```
# run the following code

from epplib.models import common
from epplibwrapper import CLIENT as registry, commands
from registrar.models.domain import Domain

domain, _ = Domain.objects.get_or_create(name="fake.gov")
domain._get_or_create_domain()
domain.expiration_date
# this should print out current expiration date

domain.renew_domain()
domain.expiration_date
# this should print out new expiration date (verifying date set in registrar)
domain.registry_expiration_date
# this should print out the new expiration date (verifying date set in cache)
```

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [x] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

**Review by Katherine-Osos — APPROVED** [review]

**Review by therealslimhsiehdy — COMMENTED** [review]
Missing context, so just want to clarify before I approve:

In Django admin as an admin, I wasn't able to edit any of the expiration dates. I also couldn't find any that were of the current date (all were expired or had `-` which I assume means they don't have one yet) so maybe that's why I couldn't edit it, but my assumption was that if they're in Ready state, that I should be able to select a date if I there wasn't one previously, or if it was expired? 😅 

Screenshot 2023-11-17 at 2 03 09 PM

Question: is it supposed to auto renew 2 years length of time?
Screenshot 2023-11-17 at 2 20 06 PM

File: src/registrar/models/domain.py [hunk 1]
```diff
@@ -211,12 +211,56 @@ def last_updated_date(self) -> date:
...
+        Renew the domain to a length and unit of time relative to the current
+        expiration date.
+
+        Default length and unit of time are 1 year.
+        """
+        # if no expiration date from registry, set to today
+        try:
+            cur_exp_date = self.registry_expiration_date
+        except KeyError:
+            logger.warning("current expiration date not set; setting to today")
```
[the remark this answers is not in the document]
Comment (abroddrick) [reply src/registrar/models/domain.py:244]: It's the date to renew on basically. So you pass the current date and then the time period of 1 year and you get "renew for up to one year from today)

**Comment by dave-kennedy-ecs** [comment]
> Missing context, so just want to clarify before I approve:
> 
> In Django admin as an admin, I wasn't able to edit any of the expiration dates. I also couldn't find any that were of the current date (all were expired or had `-` which I assume means they don't have one yet) so maybe that's why I couldn't edit it, but my assumption was that if they're in Ready state, that I should be able to select a date if I there wasn't one previously, or if it was expired? 😅
> 
> Screenshot 2023-11-17 at 2 03 09 PM
> Question: is it supposed to auto renew 2 years length of time? Screenshot 2023-11-17 at 2 20 06 PM

@therealslimhsiehdy it renews (by default) 1 year length of time.  In the screenshot, renew_domain() is called twice, hence the 2 year renewal instead of 1 year

**Comment by therealslimhsiehdy** [comment]
> > Missing context, so just want to clarify before I approve:
> > In Django admin as an admin, I wasn't able to edit any of the expiration dates. I also couldn't find any that were of the current date (all were expired or had `-` which I assume means they don't have one yet) so maybe that's why I couldn't edit it, but my assumption was that if they're in Ready state, that I should be able to select a date if I there wasn't one previously, or if it was expired? 😅
> > Screenshot 2023-11-17 at 2 03 09 PM
> > Question: is it supposed to auto renew 2 years length of time? Screenshot 2023-11-17 at 2 20 06 PM
> 
> @therealslimhsiehdy it renews (by default) 1 year length of time. In the screenshot, renew_domain() is called twice, hence the 2 year renewal instead of 1 year

Great catch! I see now that it was a "WARNING" but still did the action which is why it did 2 years :) 

Also re the "why am I only seeing `-` for expiration dates and can't edit" question: confirmed with Dave that for now the code just needs to be a getter/setter and the expiration date will just be read only for now :)

**Review by therealslimhsiehdy — APPROVED** [review]

**Pull request merged**